src/thumbnailer_utils.cc

```diff
@@ -0,0 +1,193 @@
+  std::unique_ptr<WebPAnimDecoder, void (*)(WebPAnimDecoder*)> dec(
+      WebPAnimDecoderNew(webp_data, NULL), WebPAnimDecoderDelete);
+  if (dec == NULL) {
+    std::cerr << "Error parsing image." << std::endl;
+    return kMemoryError;
+  }
+
+  WebPAnimInfo anim_info;
+  if (!WebPAnimDecoderGetInfo(dec.get(), &anim_info)) {
+    std::cerr << "Error getting global info about the animation";
+    return kGenericError;
+  }
+
+  const int width = anim_info.canvas_width;
+  const int height = anim_info.canvas_height;
+
+  while (WebPAnimDecoderHasMoreFrames(dec.get())) {
+    uint8_t* frame_rgba;
+    int timestamp;
+    if (!WebPAnimDecoderGetNext(dec.get(), &frame_rgba, &timestamp)) {
+      std::cerr << "Error decoding frame." << std::endl;
+      return kMemoryError;
+    }
+    pics->emplace_back(new WebPPicture, WebPPictureFree);
+    WebPPicture* pic = pics->back().get();
+    if (!WebPPictureInit(pic)) return kMemoryError;
+    pic->use_argb = 1;
+    pic->width = width;
+    pic->height = height;
+    if (!WebPPictureImportRGBA(pic, frame_rgba, width * 4)) {
+      return kMemoryError;
+    }
+  }
+  return kOk;
+}
+
+UtilsStatus AnimData2PSNR(const std::vector<EnclosedWebPPicture>& original_pics,
+                          WebPData* const webp_data,
+                          ThumbnailStatsPSNR* const stats) {
+  if (stats == NULL) return kMemoryError;
+
+  std::vector<EnclosedWebPPicture> pics;
+  const UtilsStatus data2pics_status = AnimData2Pictures(webp_data, &pics);
+  if (data2pics_status != kOk) return data2pics_status;
+
+  if (pics.size() != original_pics.size()) {
+    std::cerr << "Picture count mismatched.";
+    return kGenericError;
+  }
+
+  const int pic_count = original_pics.size();
+
+  for (int i = 0; i < pic_count; ++i) {
+    const EnclosedWebPPicture& orig_pic = original_pics[i];
+    const EnclosedWebPPicture& pic = pics[i];
+    float distortion_result[5];
+    if (!WebPPictureDistortion(orig_pic.get(), pic.get(), 0,
+                               distortion_result)) {
+      break;
+    } else {
+      stats->psnr.push_back(distortion_result[4]);  // PSNR-all.
+    }
+  }
+
+  // Not all frames are processed, as WebPPictureDistortion failed.
+  if (stats->psnr.size() != pic_count) return kMemoryError;
+
+  // Record statistics.
+  std::vector<float> new_psnr(stats->psnr);
+  std::sort(new_psnr.begin(), new_psnr.end());
+  stats->min_psnr = new_psnr.front();
+  stats->max_psnr = new_psnr.back();
+  stats->mean_psnr =
+      std::accumulate(new_psnr.begin(), new_psnr.end(), 0.0) / pic_count;
+  stats->median_psnr = new_psnr[pic_count / 2];
+
+  return kOk;
+}
+
+UtilsStatus CompareThumbnail(
+    const std::vector<EnclosedWebPPicture>& original_pics,
+    WebPData* const webp_data_1, WebPData* const webp_data_2,
+    ThumbnailDiffPSNR* const diff) {
+  if (original_pics.empty()) {
+    std::cerr << "Thumbnail doesn't contain any frames.";
+    return kOk;
+  }
+
+  if (diff == NULL) return kMemoryError;
```

Comment:
Use `NULL` only when dealing with C API (`WebPAnimDecoderNew()` etc.) and `nullptr` everywhere else.